<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="generator" content="VNote">

    <title>018-异常处理</title>
    <link rel="icon" href="https://github.com/tamlok/vnote/raw/master/src/resources/icons/vnote.ico">

    <style type="text/css">
    body { background-color: transparent !important; }

    </style>

    <style type="text/css">
    /* STYLE_OUTLINE_PLACE_HOLDER */
    /* STYLE_PLACE_HOLDER */
    </style>

    <!-- EXTRA_PLACE_HOLDER -->

<!-- HEAD_PLACE_HOLDER -->
</head>
<body>
<div class="container-fluid">
<div class="row flex-xl-nowrap">
    <div id="outline-panel" style="display:none;" class="d-none d-md-block d-xl-block col-md-3 col-xl-2 bd-toc">
        <div id="outline-content" class="section-nav"></div>
    </div>
    <div id="post-content" class="col-12 col-md-9 col-xl-10 py-md-3 pl-md-5 bd-content">
    <p>每一段程序都有可能出错！这是软件业的一个不容置疑的现象和规律。事实上，传统的 if…else…结构完全可以解决所有的错误，使用 Exception 机制也没能够回避在最原始的层次，通过遍历可能的情况来产生异常的做法，那么，为什么还需要异常机制</p>
<p>答案很简单：异常提供了一种更加灵活和开放的方式，使得后来的编程者可以来根据实际的情况处理这种错误，而不是使用预先设定好的处理结果。实际上，这就是异常处理机制的核心</p>
<h2 id="toc_0">使用场景<a class="vnote-anchor" href="#toc_0" data-anchor-icon="#"></a></h2>
<p>异常信息的处理方法（写日志、抛出异常），函数或过程出现异常时，只能抛出异常，禁止弹出对话框架。</p>
<p>函数及过程、控件中发生异常，通常是直接抛出异常，不要显示信息提示框；</p>
<p>界面操作（按钮）中的异常，可以显示提示信息（视具体应用而定），也可以不显示提示而将异常信息保存到日志文件或两种方式同时使用；</p>
<p>DLL文件中的异常：如果是函数或过程，发生异常时就直接抛出异常；如果是界面操作，则按界面操作（按钮）中的异常处理方法处理。</p>
<h2 id="toc_1">设计期异常<a class="vnote-anchor" href="#toc_1" data-anchor-icon="#"></a></h2>
<p>这种异常类型发生在设计期，通常是因为给组件的某个属性输入了非法的值。例如，在设计数据库应用程序时指定了一个没有定义的数据库别名。</p>
<p>这种类型的异常比较容易被发现和纠正，因为Delphi能够对属性的值进行合法性检查。一旦发现这种错误，Delphi将弹出一个警告窗口，提示你纠正错误。</p>
<h2 id="toc_2">编译期异常<a class="vnote-anchor" href="#toc_2" data-anchor-icon="#"></a></h2>
<p>编译期异常也叫语法错误，当程序代码违反了Object  Pascal的语法规则时将发生这种错误。换一种更为直白的说法就是代码写错了</p>
<p>如果程序代码中有语法错误，编译就不能通过，代码编辑器的状态栏将给出错误信息提示，并在代码编辑器中突出显示有语法错误的行。</p>
<p>比较常见的语法错误是数据类型不匹配，特别是调用RTL、VCL或Windows的API时容易发生参数不匹配的错误。</p>
<h2 id="toc_3">运行期异常<a class="vnote-anchor" href="#toc_3" data-anchor-icon="#"></a></h2>
<p>程序虽然通过了编译，但在运行时失败了，这种错误称为运行期异常</p>
<p>这种异常类型通常情况下不容易被肉眼看出来，例如，程序试图打开一个不存在的文件，在运算时出现了被零除、使用一个空对象调用了该类的成员。</p>
<h2 id="toc_4">相关语法<a class="vnote-anchor" href="#toc_4" data-anchor-icon="#"></a></h2>
<p>Delphi中异常的基类是System.SysUtils.Exception，换句话说就是所有的异常类都是Exception的子类。参考Delphi这种异常分类处理方式，我们也可以从Exception派生一个自定义的异常类</p>
<p>有人可能觉得没有自己定义异常类的必要，其实不然，就算Delphi定义的类型够多它也无法囊括所有的异常类型，而针对一些特定的场景则需要我们自己定义异常类</p>
<h3 id="toc_5">raise<a class="vnote-anchor" href="#toc_5" data-anchor-icon="#"></a></h3>
<p>在讲如何处理异常之前我们需要先了解异常是如何产生的，很简单异常的处理过程就类似一个推卸责任的过程。。。</p>
<p>raise这个关键字的主要作用是抛出一个携带异常信息的对象，而这个对象所对应的类必须是Exception或者其子类，如果在抛的过程中没有遇到处理异常的程序会一直到系统为止，具体的语法如下</p>
<pre><code class="lang-pascal hljs">
<span class="hljs-keyword">raise</span> Exception.Create(<span class="hljs-string">'我抛出的异常!'</span>);

</code></pre>
<h3 id="toc_6">try…except…end<a class="vnote-anchor" href="#toc_6" data-anchor-icon="#"></a></h3>
<p>在try 体内的代码发生异常时，系统将转向except 部分进行异常的处理。这是Delphi处理异常的最基本的方式之一。</p>
<p>只有当try 体内的代码发生异常时，才会跳转到except 里面的代码进行执行</p>
<p>需要注意的是在excpt...end之间的代码依然有可能存在异常</p>
<h3 id="toc_7">try…finally…end<a class="vnote-anchor" href="#toc_7" data-anchor-icon="#"></a></h3>
<p>这种异常处理结构一般用于保护windows的资源分配等方面，它确保了无论try 体内的代码是否发生异常，都需要由系统进行最后的统一处理的一些Windows对象的正确处理。和try…except…end不同，该结构的finally部分总被执行</p>
<blockquote>
<p>需要注意的是：不存在try…except…finally…end 结构 。其实我个人而言更喜欢这种结构，但是在Delphi中不存在这种结构</p>
</blockquote>
<h2 id="toc_8">自定义一个异常<a class="vnote-anchor" href="#toc_8" data-anchor-icon="#"></a></h2>
<p>根据前面提到的规则，定义一个异常就是定义一个Exception类的派生类，所以我们可以得到下面的代码</p>
<pre><code class="lang-pascal hljs"><span class="hljs-keyword">type</span>
    <span class="hljs-title">TArithmeticException</span> = <span class="hljs-keyword">class</span>(Exception)
    <span class="hljs-keyword">public</span>
        <span class="hljs-function"><span class="hljs-keyword">constructor</span> <span class="hljs-title">Create</span><span class="hljs-params">(Msg: <span class="hljs-keyword">string</span>)</span>;</span> <span class="hljs-keyword">overload</span>;

    <span class="hljs-keyword">end</span>;


<span class="hljs-function"><span class="hljs-keyword">constructor</span> <span class="hljs-title">TArithmeticException</span>.<span class="hljs-title">Create</span><span class="hljs-params">(Msg: <span class="hljs-keyword">string</span>)</span>;</span>
<span class="hljs-keyword">begin</span>
    <span class="hljs-keyword">inherited</span> Create(Msg);
<span class="hljs-keyword">end</span>;

</code></pre>
<p>当然，基类可以是Exception或者是Exception的任何一个任何层次的派生类</p>
<blockquote>
<p>其实构造方法可以不进行重写，但是为了代码更为规范和较高的可读性建议还是写上比较好</p>
</blockquote>
<h3 id="toc_9">抛出异常<a class="vnote-anchor" href="#toc_9" data-anchor-icon="#"></a></h3>
<p>根据不同的情况抛出异常是使用异常的最基本的模式。我的理解可能不是特别准确，所谓的抛出异常其实就是声明一个异常对象</p>
<p>例1</p>
<pre><code class="lang-pascal hljs"><span class="hljs-keyword">begin</span>
  <span class="hljs-keyword">raise</span> TArithmeticException.Create(<span class="hljs-string">'抛出异常'</span>);
<span class="hljs-keyword">end</span>;
</code></pre>
<p>例2</p>
<pre><code class="lang-pascal hljs"><span class="hljs-keyword">begin</span>
   <span class="hljs-comment">{此处因为本类没有声明这种格式的函数，所以默认引用的是父类的构造函数}</span>
  <span class="hljs-keyword">raise</span> TArithmeticException.CreateFmt(<span class="hljs-string">'%s %d'</span>, [<span class="hljs-string">'错误代码:'</span>, <span class="hljs-number">999</span>]);
<span class="hljs-keyword">end</span>;
</code></pre>
<p>例3</p>
<pre><code class="lang-pascal hljs"><span class="hljs-keyword">var</span>
  exc: Exception;
<span class="hljs-keyword">begin</span>
    <span class="hljs-comment">{这种方式和例1没有区别，仅仅是多了一个变量的声明}</span>
  exc := TArithmeticException.Create(<span class="hljs-string">'发现异常'</span>);
  <span class="hljs-keyword">raise</span> exc;
<span class="hljs-keyword">end</span>;
</code></pre>
<h2 id="toc_10">异常类型的精确处理<a class="vnote-anchor" href="#toc_10" data-anchor-icon="#"></a></h2>
<p>可以使用 on E:异常类 do... 结构可以在 do 体内处理特定异常类所抛出的异常。</p>
<p>如果在except中使用 on E: 异常类 do… 的话，在except中的 on E: 异常类 do…之外不能有任何语句，例如下面的语法是正确的</p>
<pre><code class="lang-pascal hljs"><span class="hljs-keyword">try</span>
    ...
<span class="hljs-keyword">except</span>
    <span class="hljs-keyword">on</span> E: Exception <span class="hljs-keyword">do</span>
    <span class="hljs-keyword">begin</span>
        ShowMessage(<span class="hljs-string">'OK'</span>);
        ShowMessage(<span class="hljs-string">'OK Again'</span>);
    <span class="hljs-keyword">end</span>;
<span class="hljs-keyword">end</span>;
</code></pre>
<p>但是下面的方式就错了</p>
<pre><code class="lang-pascal hljs"><span class="hljs-keyword">try</span>
    ...
<span class="hljs-keyword">except</span>
    <span class="hljs-keyword">on</span> E: Exception <span class="hljs-keyword">do</span>
    <span class="hljs-keyword">begin</span>
        ShowMessage(<span class="hljs-string">'OK'</span>);
    <span class="hljs-keyword">end</span>;
        <span class="hljs-comment">//如果在except中使用 on E: 异常类 do 的话，在except中的 on E: 异常类 do之外不能有任何语句</span>
        ShowMessage(<span class="hljs-string">'Not OK'</span>);
<span class="hljs-keyword">end</span>;
</code></pre>
<p>整个的异常处理代码</p>
<pre><code class="lang-delphi hljs"><span class="hljs-function"><span class="hljs-keyword">procedure</span> <span class="hljs-title">Demo1</span><span class="hljs-params">()</span>;</span>
<span class="hljs-keyword">begin</span>
    <span class="hljs-keyword">raise</span> TArithmeticException.Create(<span class="hljs-string">'算术异常'</span>);
<span class="hljs-keyword">end</span>;

<span class="hljs-keyword">begin</span>
    <span class="hljs-keyword">try</span>
        <span class="hljs-keyword">try</span>
            Demo1();
        <span class="hljs-keyword">except</span>
            <span class="hljs-keyword">on</span> E: TArithmeticException <span class="hljs-keyword">do</span> <span class="hljs-keyword">begin</span>

                <span class="hljs-comment">{Message即为我们前面创建对象时赋值的内容}</span>
                Writeln(E.UnitName,E.ClassName, e.<span class="hljs-keyword">Message</span>);
            <span class="hljs-keyword">end</span>;
            <span class="hljs-keyword">on</span> E: Exception <span class="hljs-keyword">do</span> <span class="hljs-keyword">begin</span>
                Writeln(<span class="hljs-string">'未知异常'</span>);
            <span class="hljs-keyword">end</span>;
        <span class="hljs-keyword">end</span>;
    <span class="hljs-keyword">finally</span>
        <span class="hljs-comment">// 此处作为资源释放的代码</span>
    <span class="hljs-keyword">end</span>;

<span class="hljs-keyword">end</span>.
</code></pre>
<blockquote>
<p>通过最后的代码可以发现，我们对于代码异常情况的处理更为详细，一旦程序出错更便于我们进行定位。</p>
</blockquote>

    </div>
</div>
</div>

<div id="container-floating" style="display:none;" class="d-none d-md-block d-xl-block">
    <div id="floating-button" onclick="toggleMore()">
        <p id="floating-more" class="more">&gt;</p>
    </div>
</div>

<!--
<div class="footer" id="vnote-footer">
    <p>Generated by <em><a href="https://tamlok.github.io/vnote/">VNote</a></em>.</p>
</div>
-->
</body>
</html>
